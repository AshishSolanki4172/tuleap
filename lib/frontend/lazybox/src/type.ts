/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { HTMLTemplateResult, TemplateResult } from "lit/html.js";
import type { GroupCollection, LazyboxItem } from "./items/GroupCollection";
import type { HTMLTemplateStringProcessor } from "./index";

export interface Lazybox {
    setDropdownContent: (groups: GroupCollection) => void;
    resetSelection: () => void;
    setSelection: (selection: ReadonlyArray<LazyboxItem>) => void;
    destroy: () => void;
}

export type LazyboxSearchFieldCallback = (query: string) => void;

export type LazyboxSelectionCallback = (selected_value: unknown | null) => void;

export type LazyboxTemplatingCallback = (
    html: typeof HTMLTemplateStringProcessor,
    item: LazyboxItem
) => HTMLTemplateResult;

export type LazyboxNewItemCallback = () => void;

export type LazyBoxWithNewItemButton = {
    readonly new_item_button_label: string;
    readonly new_item_callback: LazyboxNewItemCallback;
};

type LazyBoxWithoutNewItemButton = {
    readonly new_item_callback?: undefined;
};

export type LazyboxOptions = (LazyBoxWithoutNewItemButton | LazyBoxWithNewItemButton) & {
    readonly placeholder: string;
    readonly search_input_placeholder: string;
    readonly is_multiple: boolean;
    readonly templating_callback: LazyboxTemplatingCallback;
    readonly selection_callback: LazyboxSelectionCallback;
    readonly search_field_callback: LazyboxSearchFieldCallback;
};

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: unknown;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
}

export interface LazyboxComponent {
    wrapper_element: HTMLElement;
    lazybox_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LazyboxSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface LazyboxSelectionStateMultiple {
    selected_items: Map<string, RenderedItem>;
    selected_values_elements: Map<string, Element>;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}

export interface ManageSelection {
    processSelection(item: Element): void;
    hasSelection(): boolean;
    updateSelectionAfterDropdownContentChange(): void;
    clearSelection(): void;
    setSelection(selection: ReadonlyArray<RenderedItem>): void;
}
