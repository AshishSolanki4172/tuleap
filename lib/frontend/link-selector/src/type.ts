/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { TemplateResult } from "lit/html.js";
import type { LinkSelectorSearchFieldCallback } from "./events/SearchFieldEventCallbackHandler";

export interface LinkSelector {
    destroy: () => void;
}

export interface LinkSelectorOptions {
    readonly placeholder?: string;
    search_field_callback: LinkSelectorSearchFieldCallback;
}

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: string;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
    target_option: HTMLOptionElement;
}

export interface LinkSelectorComponent {
    wrapper_element: HTMLElement;
    link_selector_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LinkSelectorSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}
