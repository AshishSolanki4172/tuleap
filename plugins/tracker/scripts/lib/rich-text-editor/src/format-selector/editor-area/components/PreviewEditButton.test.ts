/*
 * Copyright (c) Enalean, 2021-Present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { createPreviewEditButton } from "./PreviewEditButton";
import type { GettextProvider } from "@tuleap/gettext";
import { render } from "lit-html";

describe(`PreviewEditButton`, () => {
    let mount_point: HTMLDivElement, gettext_provider: GettextProvider;
    beforeEach(() => {
        const doc = document.implementation.createHTMLDocument();
        mount_point = doc.createElement("div");
        gettext_provider = {
            gettext: (msgid): string => msgid,
        };
    });

    function getButton(): HTMLButtonElement {
        const button = mount_point.firstElementChild;
        if (!(button instanceof HTMLButtonElement)) {
            throw new Error("Could not find the button");
        }
        return button;
    }

    it.each([
        [true, "Preview", "fa-eye"],
        [false, "Edit", "fa-pencil-alt"],
    ])(
        `when is_in_edit_mode is %s, it will create a %s button`,
        (is_in_edit_mode: boolean, button_name: string, button_icon: string) => {
            const presenter = {
                is_in_edit_mode,
                onClickCallback: jest.fn(),
            };
            const template = createPreviewEditButton(presenter, gettext_provider);
            render(template, mount_point);

            const button = getButton();
            expect(button.outerHTML).toMatchSnapshot();

            const icon = button.querySelector("[data-test=button-icon]");
            if (!(icon instanceof HTMLElement)) {
                throw new Error("Could not get the button icon");
            }
            expect(icon.classList.contains(button_icon)).toBe(true);

            button.click();
            expect(presenter.onClickCallback).toHaveBeenCalled();
        }
    );
});
