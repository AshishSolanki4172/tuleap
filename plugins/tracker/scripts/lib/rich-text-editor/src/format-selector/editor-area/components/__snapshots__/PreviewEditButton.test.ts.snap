// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewEditButton if the promise fails, it will re-enable the button to allow people to retry 1`] = `
<button type="button" class="btn btn-small rte-button">
  <i aria-hidden="true" class="fas fa-fw fa-eye"></i>
  Preview
</button>
`;

exports[`PreviewEditButton when is_in_edit_mode is false, it will create a Edit button 1`] = `
<button type="button" class="btn btn-small rte-button">
  <i aria-hidden="true" class="fas fa-fw fa-pencil-alt"></i>
  Edit
</button>
`;

exports[`PreviewEditButton when is_in_edit_mode is true, it will create a Preview button 1`] = `
<button type="button" class="btn btn-small rte-button">
  <i aria-hidden="true" class="fas fa-fw fa-eye"></i>
  Preview
</button>
`;

exports[`PreviewEditButton while promise_of_preview is unsettled, it will create a disabled loading button 1`] = `
<button type="button" class="btn btn-small rte-button" disabled="">
  <i class="fas fa-fw fa-spin fa-circle-notch" aria-hidden="true"></i>
  Preview
</button>
`;
